### PR TITLE
feat: forge test suite v2 — bun:test migration, 222 new tests, gh-aw behavioral workflow

### PR DESCRIPTION
**Pull request by harshanandak**
## Summary

- **Deleted stale research lib**: Removed `lib/commands/research.js` (stub returning empty arrays), all `forge research` CLI registration, and 3 dead OpenSpec functions from `lib/commands/plan.js`
- **Migrated 41 test files** from `node:test` + `assert` → `bun:test` + `expect()` (1,862 → 2,217 expect() calls)
- **Added 222 new tests** across 6 areas: plan.js Phase 1/2/3 coverage, dev.js decision gate scoring, commitlint script, structural command-file tests, behavioral judge script, agentic workflow sync
- **Added `calculateDecisionRoute()`** to `lib/commands/dev.js` — decision gate logic now has a JavaScript implementation (was documentation-only)
- **Created gh-aw behavioral test workflow** — weekly AI-powered compliance test using real Claude runs with GLM-5/MiniMax/Kimi fallback scoring chain
- **Fixed `package.json` test script** — was an explicit file list missing all new directories; simplified to `bun test` (recursive discovery, no glob issues on Windows)

## Design Doc
See: [docs/plans/2026-02-27-forge-test-suite-v2-design.md](docs/plans/2026-02-27-forge-test-suite-v2-design.md)

## Decisions Log
See: [docs/plans/2026-02-27-forge-test-suite-v2-decisions.md](docs/plans/2026-02-27-forge-test-suite-v2-decisions.md)

Decision 1 (Task 1): `lib/commands/research.js` had live CLI reference → confirmed stub (empty arrays) → full deletion approved by user.

## Beads Issue
Closes: forge-5vf

## Key Decisions

1. **GLM-5 as primary judge model** (reasoning DISABLED): #64/66 on verbosity — must set `"reasoning": {"enabled": false}` + `response_format: json_object` to prevent score inflation. MiniMax M2.5 fallback, Kimi K2.5 last resort.
2. **3-layer scoring architecture**: Layer 1 blockers (auto-fail), Layer 2 weighted dimensions (Security ×3, TDD ×3, Design ×2, Structural ×1, max 45), Layer 3 trend tracking per-dimension.
3. **16 loophole fixes** incorporated: section existence ≠ content (min 200 chars), majority TDD threshold (>50%), file recency, placeholder detection, temperature=0 variance averaging, 4-run calibration mode, rotating test prompts, per-dimension trend alerts, cold start warmup, repo-committed scores, `workflow_run` two-workflow trigger, adversarial negative-path test, `.md`/`.lock.yml` sync check, INCONCLUSIVE for API failures.
4. **gh-aw two-workflow approach**: `detect-command-file-changes.yml` (standard YAML, watches `.claude/commands/`) triggers `behavioral-test.md` via `workflow_run` — gh-aw doesn't support direct file-path triggers.
5. **`bun test` with no arguments**: Windows glob expansion was silently failing the old explicit file list; bare `bun test` uses Bun's recursive discovery correctly.

## TDD Test Coverage

| File | Tests Added | What |
|---|---|---|
| `test/commands/plan.phases.test.js` | +71 | Phase 1/2/3 coverage: slug validation, scope detection, decision extraction, task extraction, branch/Beads/executePlan validation |
| `test/commands/dev.test.js` | +18 | Decision gate: score routing (0-3/4-7/8+), security mandatory override, `DECISION_ROUTES` constants |
| `test/scripts/commitlint.test.js` | +32 | Valid types, invalid formats, edge cases, script structure |
| `test/structural/command-files.test.js` | +51 | All 11 `.claude/commands/*.md` files: exist, non-empty, balanced code blocks, HARD-GATE count, required sections |
| `test/scripts/behavioral-judge.test.js` | +41 | GLM-5/MiniMax/Kimi fallback chain, INCONCLUSIVE path, score classification, mock curl via env vars |
| `test/structural/agentic-workflow-sync.test.js` | +9 | gh-aw `.md`/`.lock.yml` sync, CI workflow exists, model consistency |

**Final: 1215 pass, 31 skip, 0 fail** (31 skips are pre-existing — require real `gh`/`bd` binaries or Windows admin)

## Security Review

- **A03 Injection**: No user input processed — all new code is test files and CI scripts
- **A06 Vulnerable Components**: No new runtime dependencies added
- **A08 Data Integrity**: `behavioral-test-scores.json` written via `GH_AW_CI_TRIGGER_TOKEN` PAT (auth required, not unauthenticated write)
- **A09 Logging**: Behavioral test emits structured summary log per run
- All other OWASP Top 10 categories: N/A for test suite + CI workflow additions

## Test Plan
- [x] Type check: No TypeScript — skipped (exit 0)
- [x] Lint: 0 errors, 0 warnings (`bun run lint`)
- [x] Tests: 1215 pass, 31 skip, 0 fail (`bun test`)
- [x] Security review: OWASP Top 10 reviewed, no critical issues
- [x] Pre-push hooks: branch-protection ✓, lint ✓, tests ✓

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details><summary><h3>Greptile Summary</h3></summary>

This PR delivers a comprehensive test suite overhaul with 222 new tests, migrates 41 test files from node:test to bun:test, and introduces a sophisticated AI-powered behavioral compliance workflow.

**Key Changes:**
- **Test Migration**: All 41 test files migrated to `bun:test` with `expect()` assertions (1,862 → 2,217 expect() calls). Migration is consistent and thorough.
- **New Test Coverage**: 222 tests added across plan.js Phase 1/2/3 (71), dev.js decision gate (18), commitlint script (32), structural command files (51), behavioral judge (41), and agentic workflow sync (9)
- **Decision Gate Implementation**: `calculateDecisionRoute()` function added to `dev.js` with 7-dimension scoring rubric, security mandatory override, and score/dimensions drift prevention
- **Behavioral Test Workflow**: New gh-aw workflow with GLM-5/MiniMax/Kimi fallback chain, 3-layer scoring (Security ×3, TDD ×3, Design ×2, Structural ×1), 16 loophole fixes, cold-start warmup, and trend tracking
- **Stale Code Removal**: Deleted `lib/commands/research.js` (364-line stub), removed all OpenSpec functions from `plan.js` (236 lines), cleaned all CLI references
- **Test Script Fix**: Changed `package.json` test script from explicit file list to `bun test` for recursive discovery (fixes Windows glob issues)

**Previous Review Feedback:**
- ✅ `check-lock-sync` subcommand implemented (lines 21-46 in behavioral-judge.sh)
- ✅ Dead `sys.stdin.read()` default fixed (line 76 now uses empty string)
- ⚠️ Score/dimensions drift partially addressed — `effectiveScore` now derived from dimensions for routing, but mandatoryOverride case (line 482) still returns original `score` parameter. This is a minor inconsistency since route is always BLOCKED when override triggers.

**Architecture Decisions:**
- Two-workflow trigger approach for gh-aw (standard YAML workflow → `workflow_run` → agentic workflow) required because gh-aw doesn't support direct path-based triggers
- GLM-5 with `reasoning: {enabled: false}` prevents score inflation (#64/66 on verbosity)
- 3-run cold-start warmup period before trend alerts activate

All changes are well-tested with comprehensive coverage. The behavioral test workflow is sophisticated with proper fallback chains and loophole prevention. Code removal is clean with no dangling references.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

- This PR is safe to merge with minimal risk
- Score reflects high-quality test additions (222 new tests), consistent migration pattern across 41 files, clean removal of stale code with no dangling references, well-documented new functions, and comprehensive test coverage for all new features. Previous review feedback has been addressed or partially fixed. No security issues, no runtime logic changes beyond adding a pure function.
- No files require special attention — all changes are test additions, test migrations, or stale code removal
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| lib/commands/dev.js | Added `calculateDecisionRoute()` function with decision gate logic — well-documented, includes security override, derives effectiveScore from dimensions to prevent drift |
| lib/commands/plan.js | Removed 3 dead OpenSpec functions (`createOpenSpecProposal`, `formatProposalPRBody`, `createProposalPR`) — clean removal with no remaining references |
| scripts/behavioral-judge.sh | New behavioral judge script with GLM-5/MiniMax/Kimi fallback chain, includes `check-lock-sync` subcommand — comprehensive error handling, test mode supported |
| bin/forge-cmd.js | Removed all `research` command references (handler, validation, help text, examples) — clean removal with no dangling references |
| .github/agentic-workflows/behavioral-test.md | New gh-aw workflow for weekly behavioral compliance testing — 16 loophole fixes incorporated, 3-layer scoring architecture, cold-start warmup, trend tracking |
| test/commands/plan.phases.test.js | New test file with 71 tests covering Phase 1/2/3 gaps — slug validation, scope detection, decision extraction, task extraction, branch/Beads/executePlan validation |
| test/scripts/behavioral-judge.test.js | New test file with 41 tests for behavioral-judge.sh — GLM-5/MiniMax/Kimi fallback chain, INCONCLUSIVE path, score classification, test mode mocking |
| test/structural/command-files.test.js | New structural test file with 51 tests — validates all 11 `.claude/commands/*.md` files exist, are non-empty, have balanced code blocks, correct HARD-GATE count, required sections |

</details>

</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[PR: Test Suite v2] --> B[Test Migration]
    A --> C[New Tests]
    A --> D[New Features]
    A --> E[Cleanup]
    
    B --> B1[41 files: node:test → bun:test]
    B --> B2[assert → expect]
    B --> B3[1862 → 2217 expect calls]
    
    C --> C1[71: plan.js Phase 1/2/3]
    C --> C2[18: dev.js decision gate]
    C --> C3[32: commitlint script]
    C --> C4[51: structural command files]
    C --> C5[41: behavioral judge]
    C --> C6[9: agentic workflow sync]
    
    D --> D1[calculateDecisionRoute function]
    D1 --> D1a[7-dimension rubric]
    D1 --> D1b[Security override]
    D1 --> D1c[Score/dimensions validation]
    
    D --> D2[Behavioral Test Workflow]
    D2 --> D2a[GLM-5 primary judge]
    D2 --> D2b[MiniMax fallback]
    D2 --> D2c[Kimi last resort]
    D2 --> D2d[3-layer scoring]
    D2 --> D2e[16 loophole fixes]
    
    D --> D3[check-lock-sync subcommand]
    D3 --> D3a[MD/lock.yml model sync]
    
    E --> E1[Delete research.js stub]
    E --> E2[Remove OpenSpec functions]
    E --> E3[Clean CLI references]
    E --> E4[Simplify package.json test script]
```
</details>

<sub>Last reviewed commit: 3e1e7c0</sub>

### CODE REVIEW

File: .github/workflows/check-agentic-workflow-sync.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+          MD_MODEL=$(grep -A2 'engine:' .github/agentic-workflows/behavioral-test.md | grep 'model:' | awk '{print $2}')
+          # Extract engine.model from .lock.yml
+          LOCK_MODEL=$(grep -A2 'engine:' .github/agentic-workflows/behavioral-test.lock.yml | grep 'model:' | awk '{print $2}')
+
+          if [ "$MD_MODEL" != "$LOCK_MODEL" ]; then
+            echo "❌ Sync mismatch: .md model=$MD_MODEL but .lock.yml model=$LOCK_MODEL"
+            echo "Run: gh aw compile .github/agentic-workflows/behavioral-test.md"
+            exit 1
+          fi
+          echo "✅ .md and .lock.yml are in sync (model: $MD_MODEL)"
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply .github/workflows/check-agentic-workflow-sync.yml:35]: ✅ Fixed: Added permissions block to check-agentic-workflow-sync.yml

Changed: Added `permissions: contents: read` block at workflow level
Reason: New workflow created in Task 8 was missing explicit permissions declaration, triggering CodeQL actions/missing-workflow-permissions alert
Commit: ddf9a93

File: .github/workflows/detect-command-file-changes.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    branches: [master]
+    paths:
+      - '.claude/commands/**'
+
+jobs:
+  detect:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Command files changed
+        run: echo "Command file changes detected — triggering behavioral test via workflow_run"
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply .github/workflows/detect-command-file-changes.yml:17]: ✅ Fixed: Added permissions block to detect-command-file-changes.yml

Changed: Added `permissions: contents: read` block at workflow level
Reason: New workflow created in Task 8 was missing explicit permissions declaration, triggering CodeQL actions/missing-workflow-permissions alert
Commit: ddf9a93

File: test/skills-structure.test.js [hunk 3]
```diff
@@ -1,175 +1,138 @@
...
+      const content = fs.readFileSync(filePath, 'utf8');
+      expect(content.includes('core') || content.includes('base')).toBeTruthy();
+    });
+
+    test('sonarcloud-analysis SKILL.md references SonarCloud API', () => {
+      const filePath = path.join(ROOT, 'skills', 'sonarcloud-analysis', 'SKILL.md');
+      if (!fs.existsSync(filePath)) return;
+      const content = fs.readFileSync(filePath, 'utf8');
+      expect(content.includes('sonarcloud.io') || content.includes('SonarCloud')).toBeTruthy();
+    });
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply test/skills-structure.test.js:111]: This is a false positive. The code at this line is a test assertion checking whether a Markdown documentation file mentions the string 'sonarcloud.io' or 'SonarCloud'. This is not URL sanitization — there is no user-provided URL being processed or validated here.

The `content.includes('sonarcloud.io')` call simply checks that a static documentation file references SonarCloud. The value being checked is a file read from the local filesystem (a documentation file the test owns), not any external or user-controlled input. There is no security risk.

The CodeQL "Incomplete URL substring sanitization" rule applies to scenarios where user input is checked against URL prefixes to allow/deny access. This test has no such pattern.

File: .github/agentic-workflows/behavioral-test.md [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+> Note on trigger architecture (Loophole Fix 12): This workflow is triggered by `workflow_run`
+> from `detect-command-file-changes` (a standard GitHub Actions workflow that watches
+> `.claude/commands/**` on push to master). gh-aw does not support direct path-based push
+> triggers, so this two-workflow approach is required.
+
+Before doing anything else, verify that the compiled `.lock.yml` matches this workflow's frontmatter
+version field. The `.lock.yml` is the auditable compiled form and must stay in sync.
+
+```bash
+bash scripts/behavioral-judge.sh check-lock-sync .github/agentic-workflows/behavioral-test.md
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply .github/agentic-workflows/behavioral-test.md:49]: ✅ Fixed: Implemented check-lock-sync subcommand in behavioral-judge.sh

Changed: Added check-lock-sync subcommand (scripts/behavioral-judge.sh:21-46) that:
- Accepts a .md workflow file path as argument
- Derives the .lock.yml path (replaces .md extension)
- Compares engine.model from both files
- Outputs LOCK_IN_SYNC or LOCK_OUT_OF_SYNC (exit 0 in both cases per the workflow spec)

Tested: bash scripts/behavioral-judge.sh check-lock-sync .github/agentic-workflows/behavioral-test.md → LOCK_IN_SYNC
Commit: 12018b1

File: lib/commands/dev.js [hunk 5]
```diff
@@ -431,6 +431,69 @@ async function executeDev(featureName, options = {}) {
...
+	if (score <= 7) {
+		return { route: DECISION_ROUTES.SPEC_REVIEWER, score };
+	}
+
+	return { route: DECISION_ROUTES.BLOCKED, score };
+}
+
 module.exports = {
 	detectTDDPhase,
 	identifyFilePairs,
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/commands/dev.js:505]: ✅ Fixed: derive effectiveScore from dimensions to prevent score/dimensions drift

Changed: calculateDecisionRoute() now computes effectiveScore from dimensions.reduce() when dimensions.length === 7, falling back to the passed score only when dimensions are unavailable.
Reason: A caller passing a pre-computed score that doesn't match the actual dimensions sum would previously be silently misrouted. Using dimensions as the authoritative source eliminates this drift.
Commit: 3e1e7c0

File: scripts/behavioral-judge.sh [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+# ─── Subcommand: check-lock-sync ─────────────────────────────────────────────
+# Usage: bash behavioral-judge.sh check-lock-sync <workflow.md>
+# Compares engine.model in .md frontmatter against the compiled .lock.yml.
+# Outputs: LOCK_IN_SYNC or LOCK_OUT_OF_SYNC (exit 0 in both cases)
+if [ "$1" = "check-lock-sync" ]; then
+  MD_FILE="${2:-}"
+  if [ -z "$MD_FILE" ] || [ ! -f "$MD_FILE" ]; then
+    echo "LOCK_OUT_OF_SYNC" >&2
+    echo '{"result":"LOCK_OUT_OF_SYNC","reason":"md_file_not_found"}'
+    exit 0
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/behavioral-judge.sh:30]: ✅ Fixed: LOCK_OUT_OF_SYNC now consistently goes to stdout in all error paths

Changed: Swapped stdout/stderr for the md_file_not_found path — plain LOCK_OUT_OF_SYNC string to stdout, JSON detail to stderr. Now matches the lock-file-not-found path and the LOCK_IN_SYNC path.
Commit: 3e1e7c0

File: scripts/behavioral-judge.sh [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+if command -v python3 >/dev/null 2>&1; then
+  EXTRACTED=$(echo "$INPUT" | python3 -c "
+import sys, json
+try:
+    d = json.load(sys.stdin)
+    print(d.get('plan_output', sys.stdin.read() if hasattr(sys.stdin,'read') else ''))
+except:
+    pass
+" 2>/dev/null) || true
+  if [ -n "$EXTRACTED" ]; then
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/behavioral-judge.sh:82]: ✅ Fixed: removed dead sys.stdin.read() default in Python inline script

Changed: Replaced d.get('plan_output', sys.stdin.read() ...) with d.get('plan_output', '') and added an if-guard before printing — json.load(sys.stdin) exhausts the stream so the old default always returned empty string.
Commit: 3e1e7c0
